### DISCUSSION TITLE
proper way to use useWatch along with useFieldArray

### DISCUSSION THREAD

**Discussion by ejose19** [react-hook-form/react-hook-form#2036] [Q&A]
Hello, is there a way to use "useWatch" along with "useFieldArray" that doesn't re-render all other elements in the array when only one of it is changed/touched?

**Comment by bluebill1049** [comment]
useFieldArray can't avoid re-render when you delete insert and swap items.

**Reply by ejose19** [reply]
Hello @bluebill1049 thanks for your quick response, I'm neither adding or inserting elements, list is basically static. The issue comes when I alter individual properties of each element / or just clicking on an input for one element makes all other elements re-render onBlur

**Reply by bluebill1049** [reply]
can you share a simple codesandbox?

**Reply by ejose19** [reply]
Was digging a bit more and seems the issue is related to using "useWatch" with the array form ( name: [x, y] ), when using it individually the issue is not present. Here is the [sandbox](https://codesandbox.io/s/eloquent-dust-72pwb?file=/src/Root.tsx)

**Reply by bluebill1049** [reply]
is this still an issue with offical v6? (sorry miss the reply)

**Reply by ejose19** [reply]
Yes, sandbox was submitted with v6, updated script to 6.0.1 and issue is still present, check Mapper file on the sandbox for the details of the issue

**Reply by bluebill1049** [reply]
> Was digging a bit more and seems the issue is related to using "useWatch" with the array form ( name: [x, y] ), when using it individually the issue is not present. Here is the [sandbox](https://codesandbox.io/s/eloquent-dust-72pwb?file=/src/Root.tsx)

useWatch is made for individual fields, otherwise, you probably have to use `watch`

**Reply by ejose19** [reply]
Should the doc be updated to reflect that? I was using it as shown on TS example https://react-hook-form.com/api/#useWatch like:
```ts
const namesOnly2 = useWatch<{
  test: string;
  test1: number;
}>({ name: ['test', 'test1'] });
```

**Reply by bluebill1049** [reply]
you mean this: https://codesandbox.io/s/react-hook-form-v6-usewatch-bvevw?file=/src/index.js

**Reply by ejose19** [reply]
Yes, as you see even in that example it works good, the issue shows when using that { name: [...] } arguments along with useFieldArray

**Reply by bluebill1049** [reply]
i will do some investigation on this one.

**Reply by ejose19** [reply]
Here's another [sandbox](https://codesandbox.io/s/react-hook-form-v6-usewatch-s2fz2?file=/src/index.js) using the link you sent where I reproduced the issue as well

**Reply by bluebill1049** [reply]
can you leave the steps to reproduce as well?

**Reply by ejose19** [reply]
Create 3 components (Main, Mapper, Item).

- Main instanciates the form with defaultValues, which has a property with an array of objects that will consume useFieldArray
- Main also creates a FormProvider and inside it will be Mapper component
- Mapper components calls useWatch using the array arguments (watching unrelated fields, none coming from array)
- Mapper iterates over fields and returns Item
- Item register input using using register provided by context

Notes
- Using FormProvider doesn't change anything, so even if not using it and passing control/register as props the result is the same.
- Using useWatch in the Item component with the array args will cause siblings component to re-render (but not mapper), using "useWatch" once per field works as expected

**Reply by bluebill1049** [reply]
So the watching is working as expcted as below:

<img width="1289" alt="Screen Shot 2020-07-07 at 5 59 05 pm" src="https://user-images.githubusercontent.com/10513364/86742583-92863200-c07b-11ea-8f75-3b1b8de651a8.png">

<img width="1280" alt="Screen Shot 2020-07-07 at 6 00 47 pm" src="https://user-images.githubusercontent.com/10513364/86742878-ca8d7500-c07b-11ea-95d1-f04e29023cd1.png">

Can you point me to the steps expected behavior?

**Reply by ejose19** [reply]
> Here's another [sandbox](https://codesandbox.io/s/react-hook-form-v6-usewatch-s2fz2?file=/src/index.js) using the link you sent where I reproduced the issue as well

To notice to bug: Open the sandbox, don't change anything on the code and start typing on item 1, see on the console both items and the parent component re-rendered, that's not expected behaviour since parent is not watching items fields.

To see expected behaviour: Comment this:
```ts
useWatch({
  name: ["firstName", "lastName"]
});
```
and uncomment this:
```ts
const fn = useWatch({name: 'firstName'});
const ln = useWatch({name: 'lastName'})
```

start typing on item 1 again, you will see this time neither item re-renders (after initial 2 complete rerenders which I assume is caused by form getting dirty)

If you also enable the useWatch inside the items, the re-renders will also remain inside the component and nothing else

**Reply by bluebill1049** [reply]
Thanks for the detailed explanation! that helps, I will take a look again lunchtime today. 👍

**Reply by bluebill1049** [reply]
<img width="1272" alt="Screen Shot 2020-07-09 at 10 32 42 am" src="https://user-images.githubusercontent.com/10513364/86983564-87dcb180-c1cf-11ea-9eb3-c3eed5e369bf.png">
<img width="1366" alt="Screen Shot 2020-07-09 at 10 32 58 am" src="https://user-images.githubusercontent.com/10513364/86983582-90cd8300-c1cf-11ea-80c2-ceb0c4c4a609.png">

looks like they share the same behavior from the console log, perhaps i miss the step?

**Reply by bluebill1049** [reply]
> If you also enable the useWatch inside the items, the re-renders will also remain inside the component and nothing else

That's the intended behavior.

**Reply by bluebill1049** [reply]
also `dirty` is not involved as in the codesandbox you are not subscribed to formState, so if there is an issue, it's something else.

**Reply by ejose19** [reply]
@bluebill1049 You're typing on First Name, not on item 1 (where 'phone' is located). Go to either item 1 or 2, and type '12345'. You will notice first 'way' will stop re-render once you tap '3', but the array useWatch never stop rerendering

**Reply by ejose19** [reply]
> also `dirty` is not involved as in the codesandbox you are not subscribed to formState, so if there is an issue, it's something else.

@bluebill1049 ok, dirty was my assumption, then there's another problem as modifying an item shouldn't make parent components re-render

**Reply by ejose19** [reply]
> > If you also enable the useWatch inside the items, the re-renders will also remain inside the component and nothing else
> 
> That's the intended behavior.

Yes, this was just an observation

**Reply by bluebill1049** [reply]
> @bluebill1049 You're typing on First Name, not on item 1 (where 'phone' is located). Go to either item 1 or 2, and type '12345'. You will notice first 'way' will stop re-render once you tap '3', but the array useWatch never stop rerendering

oops, i will test again.

**Reply by bluebill1049** [reply]
hmmm good one, try is useWatch subscribe to item...

**Reply by ejose19** [reply]
So to summarize there are 2 issues:

1. useWatch in array way ( name: ['x', 'y'] ) along with useFieldArray causes the component to re-render when any form value changes/blur, even if they're not being watched
2. useWatch(both with array way or individual useWatch)  along with useFieldArray, re-renders the component at least twice when a child component (array item) changed, even when caller component is not watching any array item

**Reply by bluebill1049** [reply]
thanks, the issue is useWatch shouldn't re-render at FieldArray, i am investigating for the reason why?

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-v6-usewatch-u8ui6?file=/src/index.js
